README.md
paleofire
=========
paleofire: an R package to analyse sedimentary charcoal records from the Global Charcoal Database to reconstruct past biomass burning

The `paleofire` package provides tools to extract and analyse charcoal sedimentary data stored in the Global Charcoal Database (see http://gpwg.paleofire.org for details). Main functionalities includes data extraction and sites selection, transformation and homogenization of the charcoal records as well as regional to global compositing.


Installation:
=============

To install `paleofire` from CRAN simply type `install.packages("paleofire")` at the R prompt.  

To install `paleofire` from GitHub the `devtools` package is required: on Windows platform the Rtools.exe program is required in order to install the `devtools` package. Rtools.exe can be downloaded for a specific R version on http://cran.r-project.org/bin/windows/Rtools/

Once `devtools` is installed type the following lines at R prompt to install `paleofire`:

```R
library(devtools)
install_github(repo="oblarquez/paleofire")
# The GCD package is required along paleofire:
install_github(repo="oblarquez/GCD")
library(paleofire)
```

To test everything is working you can plot a map of all charcoal records included in the Global Charcoal Database:

```R
plot(pfSiteSel())
```

For details and examples about `paleofire` please refer to the included [manual](  https://CRAN.R-project.org/package=paleofire/paleofire.pdf) or [vignette](https://CRAN.R-project.org/package=paleofire/vignettes/paleofire-paper.pdf).

Citation:
=============

To cite `paleofire` in publications use:

  Blarquez O., Vannière B., Marlon J. R., Daniau A.-L. Power M. J. Brewer S.
  and Bartlein P. J. 2014. paleofire: An R package to analyse sedimentary
  charcoal records from the Global Charcoal Database to reconstruct past
  biomass burning. Computers and Geosciences, 72, 255-261
  
```R
citation (package="paleofire")
```

Maintainer: Olivier Blarquez <blarquez@gmail.com>


NEWS
paleofire 1.2.4 01/12/2019
-----------------------------------------

* Minor additions and graphic options,


paleofire 1.2.3 29/10/2018
-----------------------------------------

* New SEA function for superposed epoch analysis using block bootstrap

* Minor bug fixes

paleofire 1.2.2 25/04/2018
-----------------------------------------

* GCD version numbers and update dates added in GCD package to enable reproducing previous vignette example:
 ID <- pfSiteSel(lat>30 & lat<90, long>-100 & long<(-50), date_int<=2500, num_version<400) now reproduce the Computer and Geoscience paper results.

* minor bug fixes 


paleofire 1.2.1 01/02/2018
-----------------------------------------

* Support GCD package v4.0.0

* id_region codes in pfSiteSel are deprecated use "continent" instead 


paleofire 1.2.0 28/09/2017
-----------------------------------------

* Support R 3.4

* Minor bug fixes

paleofire 1.1.9 19/09/2016
-----------------------------------------

* Improved kdffreq function

* pfAddData options for csv and examples

* contrib method for pfCompositeLF objects

* Bug fixes


paleofire 1.1.8 21/12/2015
-----------------------------------------

New feature:

* Can now rbind multiple pfAddData objects

Minor fixes:

* NAMESPACE

* Various fixes for R 3.2.3


paleofire 1.1.7 05/11/2015
-----------------------------------------

Minor fixes:

* Projections (longlat usage now permitted in pfGridding)

* No transformation available in pfTransform

* Various fixes


paleofire 1.1.6 22/04/2015
-----------------------------------------

New features:

* pfSimpleGrid function for plotting simple gridded charcoal maps

* contiguous for checking whether charcoal seris are sampled contiguoulsy or not

* kdffreq function for calculating paleo fire frequencies using kernel density estimation 

Fixes:

* All doc moved to roxygen

* Potveg function moved from GCD package

* Various bug fixes


paleofire 1.1.5 25/11/2014
-----------------------------------------

* Numerous bug fixes 

* Moving of packages used in a single function to Suggests


paleofire 1.1.4 18/11/2014
-----------------------------------------

New features:

* New pfToKml function for easy producing kml files from pfSiteSel objects

* New vignette based on the Computers and Geosciences paper see citation(package="paleofire")

* Updates to pfDotMap 

Fixes:

* Minor bug fixes in pfGridding and others.


paleofire 1.1.3 22/05/2014
-----------------------------------------

New features:

* pfPublications function to get citation for charcoal series

* Option added to pfGridding for using additionnal spatailpolygons dataframe as mask in the analysis 

Fixes:

* Fixed projection and cell sizes in pfGridding and colors in associated plot

* Minor fixes for graphical outputs

paleofire 1.1.2 15/03/2014
-----------------------------------------

Fix for windows compil related to sink usage...

paleofire 1.1 14/03/2014
-----------------------------------------

New features:

* Rewritten pfSiteSel for more flexibility (use of | & > < <= >= != %in% etc.)

* Added pfKruskal for testing time periods for equality in transformed char

* Added pfDotMap and pfGridding to produce map graphics representing spatial variability in charcoal data

Minor new features:

* Renamed variables for consistency (?pfSiteSel)

* Graphics controls enhancements

* verbose options


paleofire 1.0 08/01/2014
-----------------------------------------

* Initial release of the package


R/pfKruskal.R
#' Analyse composite records by a Kruskal-Wallis ANOVA
#'
#' The function applies a Kruskal-Wallis ANOVA on binned data issued from a
#' "pfComposite" object (of directly on "pfTransform" objects), in order to
#' test the difference in biomass burning activity between different time
#' periods.
#'
#'
#' @param data An object returned by \code{\link[paleofire]{pfComposite}} or
#' \code{\link[paleofire]{pfTransform}}.
#' @param p.adj Method for adjusting p values (see
#' \code{\link[stats]{p.adjust}}). Includes: "holm", "hochberg", "hommel",
#' "bonferroni", "BH", "BY", "fdr" and "none" (default).
#' @param alpha Numeric, confidence level.
#' @param bins Numeric, bins to use if a "pfTransform" object is provided.
#' @param verbose Logical, verbose or not...
#' @return A "pfKruskal" object containing multiple comparison results.
#' @author O. Blarquez
#' @seealso \code{\link{plot.pfKruskal}},\code{\link[agricolae]{kruskal}}
#' @examples
#' \dontrun{
#' ## Composite charcoal record for Western Boreal North America:
#' ID=pfSiteSel(continent=="North America", long<(-100) & l12==1)
#' plot(ID)
#' ## Transform data
#' res3=pfTransform(ID,method=c("MinMax","Box-Cox","Z-Score"),BasePeriod=c(200,4000))
#'
#' ## Composite
#' comp=pfComposite(res3,bins=seq(from=-500,to=12500,by=1000))
#' plot(comp)
#'
#' ## Kruskal Wallis Anova
#' comparison=pfKruskal(comp)
#' }
pfKruskal <- function(data, p.adj="none",
                      alpha=0.05, bins=NULL,
                      verbose=TRUE) {
  if (!requireNamespace("agricolae", quietly = TRUE)) {
    install.packages("agricolae")
  }

  if (class(data) == "pfComposite") {
    data <- cbind(x = rep(data$BinCentres, ncol(data$BinnedData)), y = c(data$BinnedData))
    data <- data.frame(data)
  }

  if (class(data) == "pfTransform") {
    data <- data.frame(na.omit(cbind(y = c(data$TransData), x = c(data$Age))))
    if (is.null(bins)) bins <- seq(-500, 12500, 1000)
    xx <- as.numeric(cut(data$x, breaks = bins))

    hbins <- bins + max(diff(bins) / 2)
    hbins <- data.frame(cbind(1:(length(hbins) - 1), hbins[1:(length(hbins) - 1)]))
    id <- with(hbins, X2[match(xx, X1)])

    data <- data.frame(na.omit(cbind(x = id, y = data$y)))
  }

  ## Post-hoc
  data$x <- as.factor(data$x)
  comparison <- agricolae::kruskal(data$y, data$x, alpha = alpha, p.adj = p.adj, group = TRUE, main = NULL)

  ## Data prep
  comparison$groups <- data.frame(
    trt = rownames(comparison$groups), means = comparison$groups$`data$y`,
    M = comparison$groups$groups
  )

  signifL <- as.data.frame(comparison$group)
  signifL$M <- as.character(signifL$M)

  signifL$trt <- as.numeric(paste(comparison$group$trt))
  signifL <- signifL[order(signifL$trt), ]

  ## Pos of letters

  test <- boxplot(data$y ~ data$x)$stats[5, ]
  garbage <- dev.off()

  signifL <- data.frame(cbind(signifL, lpos = test))
  signifL$trt <- as.factor(signifL$trt)
  colnames(signifL) <- c("x", "means", "lab", "lpos")

  # X reverse
  data$x <- factor(data$x, levels = rev(levels(data$x)))

  if (verbose == TRUE) print(comparison)

  out <- list(letters = signifL, data = data, summary = comparison)
  class(out) <- "pfKruskal"
  return(out)
}





#' Plot a "pfKruskal" object.
#'
#' Plot a "pfKruskal" object using boxplots and showing significant differences
#' between the periods using letters.
#'
#' If two periods share the same letter their rank (median) is not
#' significantly different at the confidence level specified by \code{alpha}.
#' If not, equality could be rejected at the confidence level specified by
#' \code{alpha}.
#'
#' @method plot pfKruskal
#' @export
#' @param x An object returned by \code{\link[paleofire]{pfKruskal}}.
#' @param trend Logical, show trend using linear regression?
#' @param outliers Logical, show outliers?
#' @param xlim Numeric, x axis limits.
#' @param ylim Numeric, y axis limits.
#' @param \dots ...
#' @return Return a ggplot2 "gg" object.
#' @author O. Blarquez
#' @seealso \code{\link[paleofire]{pfKruskal}}
#' @examples
#' \dontrun{
#' ## Composite charcoal record for Western Boreal North America:
#' ID=pfSiteSel(continent=="North America", long<(-100) & l12==1)
#' plot(ID)
#' ## Transform data
#' res3=pfTransform(ID,method=c("MinMax","Box-Cox","Z-Score"),BasePeriod=c(200,4000))
#'
#' ## Composite
#' comp=pfComposite(res3,bins=seq(from=-500,to=12500,by=1000))
#' plot(comp)
#'
#' ## Kruskal Wallis Anova
#' comparison=pfKruskal(comp)
#'
#' plot(comparison)
#'
#' # p=plot(comparison)
#' # require(ggplot2)
#' # p+ggtitle("my title")
#' }
plot.pfKruskal <- function(x, trend=FALSE, outliers=FALSE, xlim = NULL, ylim = NULL, ...) {
  y <- lpos <- lab <- NULL
  x$data <- na.omit(x$data)
  ### GGplot2 boxplot
  p <- ggplot(data = x$data, aes(x = x, y = y))
  if (outliers == TRUE) p <- p + geom_boxplot(fill = "grey90", outlier.size = 1, show.legend = FALSE)
  if (outliers == FALSE) p <- p + geom_boxplot(fill = "grey90", outlier.size = 0, show.legend = FALSE)
  p <- p + theme_bw(base_size = 18) + xlab("Age (cal BP)") + ylab("Composite") +
    geom_text(aes(x = x, y = lpos + 0.1, label = lab), data = x$letters, vjust = 0, size = 5) +
    coord_cartesian(xlim = xlim, ylim = ylim)
  # geom_text(data=data.frame( x = 15, y= 2.7), map=aes(x=x, y=y), label = w)

  if (trend == TRUE) {
    # p=p+geom_text(data=data.frame( x = 1, y= 3), map=aes(x=x, y=y), label = label)
    p <- p + geom_smooth(method = "lm", se = FALSE, color = "black", aes(group = 1))
  }
  p
  return(p)
}
